be/src/exec/pipeline/aggregate/spillable_aggregate_blocking_sink_operator.cpp

```diff
@@ -77,28 +105,62 @@ Status SpillableAggregateBlockingSinkOperator::push_chunk(RuntimeState* state, c
         return Status::OK();
     }
-    RETURN_IF_ERROR(_aggregator->evaluate_groupby_exprs(chunk.get()));
-    if (_spill_strategy == spill::SpillStrategy::NO_SPILL) {
-        return AggregateBlockingSinkOperator::push_chunk(state, chunk);
-    } else if (_spill_strategy == spill::SpillStrategy::SPILL_ALL) {
+    RETURN_IF_ERROR(AggregateBlockingSinkOperator::push_chunk(state, chunk));
+    set_revocable_mem_bytes(_aggregator->hash_map_memory_usage());
+    if (_spill_strategy == spill::SpillStrategy::SPILL_ALL) {
         return _spill_all_inputs(state, chunk);
-    } else {
-        return Status::OK();
     }
+    return Status::OK();
 }
 
 Status SpillableAggregateBlockingSinkOperator::_spill_all_inputs(RuntimeState* state, const ChunkPtr& chunk) {
     // spill all data
-    auto spillable = std::make_shared<Chunk>();
-    RETURN_IF_ERROR(_aggregator->convert_to_spill_format(chunk.get(), &spillable));
+    DCHECK(!_aggregator->is_none_group_by_exprs());
+    _aggregator->hash_map_variant().visit(
+            [&](auto& hash_map_with_key) { _aggregator->it_hash() = _aggregator->_state_allocator.begin(); });
+    CHECK(!_aggregator->spill_channel()->has_task());
+    RETURN_IF_ERROR(_spill_aggregated_data(state));
+    return Status::OK();
+}
 
-    auto executor = _aggregator->spill_channel()->io_executor();
-    RETURN_IF_ERROR(
-            _aggregator->spiller()->spill(state, spillable, *executor, spill::MemTrackerGuard(tls_mem_tracker)));
+Status SpillableAggregateBlockingSinkOperator::_spill_aggregated_data(RuntimeState* state) {
+    auto io_executor = _aggregator->spill_channel()->io_executor();
+    auto spiller = _aggregator->spiller();
+    auto spill_channel = _aggregator->spill_channel();
+    auto process_task = _generate_spill_task(state);
+    while (!_aggregator->spiller()->is_full()) {
+        auto chunk_st = process_task();
+        if (chunk_st.ok()) {
+            RETURN_IF_ERROR(
+                    spiller->spill(state, chunk_st.value(), *io_executor, spill::MemTrackerGuard(tls_mem_tracker)));
+        } else if (chunk_st.status().is_end_of_file()) {
+            // no more data in aggregator, reset state
+            RETURN_IF_ERROR(_aggregator->reset_state(state, {}, nullptr));
+            return Status::OK();
+        } else {
+            return chunk_st.status();
+        }
+    }
+    spill_channel->add_spill_task(std::move(process_task));
 
     return Status::OK();
 }
 
+std::function<StatusOr<ChunkPtr>()> SpillableAggregateBlockingSinkOperator::_generate_spill_task(RuntimeState* state) {
+    return [this, state]() -> StatusOr<ChunkPtr> {
+        bool use_intermediate_as_output = true;
+        if (!_aggregator->is_ht_eos()) {
+            auto chunk = std::make_shared<Chunk>();
+            RETURN_IF_ERROR(
+                    _aggregator->convert_hash_map_to_chunk(state->chunk_size(), &chunk, &use_intermediate_as_output));
+            chunk->check_or_die();
```

Comment:
why die? DCHECK()?

Reply:
> why die? DCHECK()?

this code is for debugging, I have removed it.